src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -1946,6 +1978,41 @@ public SocketError SendFileAsync(SafeFileHandle fileHandle, long offset, long co
             return SocketError.IOPending;
         }
 
+        // Called on the epoll thread, speculatively tries to process synchronous events and errors for synchronous events, and
+        // returns any remaining events that remain to be processed. Taking a lock for each operation queue to deterministically
+        // handle synchronous events on the epoll thread seems to significantly reduce throughput in benchmarks. On the other
+        // hand, the speculative checks make it nondeterministic, where it would be possible for the epoll thread to think that
+        // the next operation in a queue is not synchronous when it is (due to a race, old caches, etc.) and cause the event to
+        // be scheduled instead. It's not functionally incorrect to schedule the release of a synchronous operation, just it may
+        // lead to thread pool starvation issues if the synchronous operations are blocking thread pool threads (typically not
+        // advised) and more threads are not immediately available to run work items that would release those operations.
+        public unsafe Interop.Sys.SocketEvents HandleSyncEventsSpeculatively(Interop.Sys.SocketEvents events)
+        {
+            if ((events & Interop.Sys.SocketEvents.Error) != 0)
+            {
+                // Set the Read and Write flags; the processing for these events
+                // will pick up the error.
+                events ^= Interop.Sys.SocketEvents.Error;
+                events |= Interop.Sys.SocketEvents.Read | Interop.Sys.SocketEvents.Write;
+            }
+
+            if ((events & Interop.Sys.SocketEvents.Read) != 0 &&
+                _receiveQueue.IsNextOperationSynchronous_Speculative &&
+                _receiveQueue.ProcessSyncEventOrGetAsyncEvent(this) == null)
```

Comment:
@kouvel @stephentoub @adamsitnik I think there is an issue when `IsNextOperationSynchronous_Speculative` is `true` but the operation is not really a SyncEvent operation. The queue moves to `Processing` but no-one dispatches the operation.

Reply:
Nice catch, will fix